### ISSUE TITLE
Adding the ability for getpass to print asterisks when password is typed

### ISSUE THREAD

**Issue by c46a4f13-986a-406e-a0ff-a46d1640a576** [python/cpython#77065]
BPO | [32884](https://bugs.python.org/issue32884)
--- | :---
Nosy | @stevendaprano, @bitdancer, @jab, @MatanyaStroh, @stevoisiak, @remilapeyre, @websurfer5, @akulakov

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-02-20.10:28:53.338>
labels = ['type-feature', 'library', '3.9']
title = 'Adding the ability for getpass to print asterisks when password is typed'
updated_at = <Date 2021-08-10.01:27:06.609>
user = 'https://github.com/MatanyaStroh'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-10.01:27:06.609>
actor = 'andrei.avk'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-02-20.10:28:53.338>
creator = 'matanya.stroh'
dependencies = []
files = []
hgrepos = []
issue_num = 32884
keywords = []
message_count = 7.0
messages = ['312410', '312520', '312745', '339803', '344784', '375038', '399298']
nosy_count = 9.0
nosy_names = ['steven.daprano', 'r.david.murray', 'jab', 'matanya.stroh', 'stevoisiak', 'remi.lapeyre', 'Jeffrey.Kintscher', 'celal.sahin', 'andrei.avk']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue32884'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-130496
* gh-133849
* gh-134053
* gh-134058
* gh-134059

**Comment by thibaudcolas** [comment]
I’m surprised to see this opened for so long! Is this waiting on a design decision or pull request? Looking at the [contribution guidelines for standard library changes](https://devguide.python.org/developer-workflow/stdlib/#stdlibchanges), it seems the recommended course of action would be to create a new thread in Discourse under [Ideas](https://discuss.python.org/c/ideas/6)?

My two cents on the issue – it’s one of those we run into at every single workshop / training for newcomers to open source like [Django Girls](https://djangogirls.org/). There‘s no feedback when typing. In the case of Django, there’s no indication not to expect any feedback either. Lots of people use Windows and will have never come across that kind of prompt ever. So as mentioned above – this is indeed encouraging people to choose a password that’s simple to type.

**Comment by Stevoisiak** [comment]
> I’m surprised to see this opened for so long! Is this waiting on a design decision or pull request? Looking at the [contribution guidelines for standard library changes](https://devguide.python.org/developer-workflow/stdlib/#stdlibchanges), it seems the recommended course of action would be to create a new thread in Discourse under [Ideas](https://discuss.python.org/c/ideas/6)?

I don't see anyone opposed to this change, but somebody would need to step up to write an implementation that works on Unix.

**Comment by thibaudcolas** [comment]
Thanks for looking into this @Stevoisiak. Would it be acceptable to fix this for Windows users to start with, and change it for Unix users later?

From my perspective I see this has been opened for close to 6 years now even though people seem to think it’s a simple change for Windows. As a regular Django Girls coach this has basically been an issue for Windows users at every event I’ve ever coached at, where we run into this when creating a user account’s password for Django.

**Comment by Stevoisiak** [comment]
> Thanks for looking into this @Stevoisiak. Would it be acceptable to fix this for Windows users to start with, and change it for Unix users later?
> 
> From my perspective I see this has been opened for close to 6 years now even though people seem to think it’s a simple change for Windows. As a regular Django Girls coach this has basically been an issue for Windows users at every event I’ve ever coached at, where we run into this when creating a user account’s password for Django.

I've also wanted to see this feature for a while, ([I even proposed a Windows implementation 4 years ago](https://github.com/python/cpython/issues/77065#issuecomment-1093777160)), but I implementing it only on Windows would likely lead to confusion. If a user writes a script with `getpass.getpass(show_asterisks=True)`, the user would reasonably expect their script to show asterisks regardless of what platform the script is being run on.

**Comment by thibaudcolas** [comment]
Isn’t it the current implementation that defies people’s expectations? The current `win_getpass` function implements the behavior of `getpass` from Unix (to log in?), which never shows passwords, so is "as-expected" for Unix users of Python I imagine. While on Windows – password prompts definitely show asterisks, hence why it’s confusing the ones implemented with Python don’t?

**Comment by mjbear** [comment]
@thibaudcolas 
As of November 2020 a package called `maskpass` exists.
(It hasn't seen a release since November 2022 so maybe this is an issue or maybe it isn't.)

https://pypi.org/project/maskpass/
https://github.com/FuturisticGoo/maskpass

Certainly there are benefits for the standard library having this functionality though. :wink:
Hope this helps.

**Comment by thibaudcolas** [comment]
omg! Yes please :)

**Comment by mjbear** [comment]
> > Thanks for looking into this @Stevoisiak. Would it be acceptable to fix this for Windows users to start with, and change it for Unix users later?
> > From my perspective I see this has been opened for close to 6 years now even though people seem to think it’s a simple change for Windows. As a regular Django Girls coach this has basically been an issue for Windows users at every event I’ve ever coached at, where we run into this when creating a user account’s password for Django.
> 
> I've also wanted to see this feature for a while. ([I proposed a Windows implementation in 2019](https://github.com/python/cpython/issues/77065#issuecomment-1093777160), which itself may have been [based on a post from 2013](https://stackoverflow.com/a/16670956/3357935)), but implementing it only on Windows would likely lead to confusion. If this were officially implemented and a Python script calls `getpass.getpass(show_asterisks=True)`, most reasonable users would expect their script to show asterisks regardless of whether it's running on Windows, Linux/Unix, Mac, or any other officially supported platform.

:100:% agree with @Stevoisiak that if the asterisks masking feature is added it should be across all platforms.

Would it be unnecessary scope creep to allow the masking character to be user-defined?
(The default mask character wouldn't do anything unless the boolean is True.)
(Checking the mask character string length < 2 as well.)

```python
getpass.getpass(show_mask=True, mask_character='*')

# or a Unicode bullet character
getpass.getpass(show_mask=True, mask_character='•')

# or even more creative with Unicode (ok, yes, getting silly here)
getpass.getpass(show_mask=True, mask_character='🐧')
```

The simplest road forward is likely to go with asterisks as a "static" mask character first.
If a configurable masking character is blessed/allowed then make that phase two of changes.

**Comment by donBarbos** [comment]
looks like a good idea, I'm ready to start working on it and I would also like to suggest an idea to replace:
```python
getpass.getpass(show_mask=True, mask_character='*')
```
to:
```python
getpass.getpass(mask_character='*')
```
with mask_character=None by default, and it will be displayed if bool(mask_character) == True (`if mask_character: ...`)

I also want to ask are you sure that we need a check for the length of `mask_character`, because some emoji may have a longer length or the user may want to use their own custom string. I'm not insisting, I just think this restriction is far-fetched.

**Comment by mjbear** [comment]
> to:
> 
> getpass.getpass(mask_character='*')
> 
> with mask_character=None by default, and it will be displayed if bool(mask_character) == True (`if mask_character: ...`)

I like the idea to default to None.

And yes - it can simply be:

```python
if mask_character:
    # display mask character to stdout
```
 
> I also want to ask are you sure that we need a check for the length of `mask_character`, because some emoji may have a longer length or the user may want to use their own custom string. I don't insist, I just think this restriction is far-fetched.

@donBarbos I wasn't thinking about multi-character emojis. :wink:

Either way, since you bring it up, it seems reasonable to allow for flexibility. :+1:

**Comment by JelleZijlstra** [comment]
Looks like a PR was merged for this, so this could be closed.

However, there's an issue: the PR added the new parameter only to `unix_getpass` and `win_getpass` but not to `fallback_getpass`. Let's add it there too for consistency.

**Comment by picnixz** [comment]
Right, it's also needed at runtime as `getpass` is set to `fallback_getpass` in case msvcrt fails to be imported.

**Comment by donBarbos** [comment]
I can send PR soon.
note: I left a comment about this https://github.com/python/cpython/pull/130496#issuecomment-2692358030
